### ISSUE TITLE
Elasticsearch version 5.x support

### ISSUE THREAD

**Issue by vascop** [datarank/tempest#24]
According to the readme, this is plugin is only supporting ES up to version 2.3.2. Are there plans for 5.x (5.6) support? If not, could you offer some pointers on how to start if one wanted to contribute support for more recent versions?

**Comment by sukantasaha** [comment]
Any update on this , we need this plugin for ES 5.X

**Comment by kennycason** [comment]
@sukantasaha @vascop I just pm'd team and will discuss soon.

**Comment by Allen704** [comment]
I'm assuming at this point the answer is "no".

**Comment by aewhite** [comment]
Sorry, I'm just now seeing this. We are currently evaluating ES 5 and 6 but it's not a high priority yet. I have attempted to compile this against ES 5 and the results indicated that a rather large API change had taken place.

**Comment by aewhite** [comment]
It looks like nearly every integration point broke between 2.x and 5.x. I'll see what I can do but it probably won't happen for a while. Especially since I don't have a good 5.x cluster to test on.

**Comment by reedflinch** [comment]
After upgrading from 2.4 to 5.6 we'd LOVE to be able to use this. If anyone has any other resources to balance shards across nodes by size, I'd be very interested in hearing about them.

**Comment by aewhite** [comment]
So, I'm working on a 5.6 branch now see https://github.com/datarank/tempest/tree/v2.2.2-ES5.6.6-SNAPSHOT for details. I still have some compile-time/interface issues to resolve and then I have to test it. I think I can have something out in a few days but it's hard to say for sure.

**Comment by muffix** [comment]
@aewhite Do you have any update on the state of 5.x (or higher) support? Anything you could use a hand with?

**Comment by luflow** [comment]
@aewhite Is there any chance to support a further development on this? We also are in need of size aware rebalancing. Especially for 7.x

Please let me know if we can support you with his. E.g. by sponsoring or conributing to the development!

**Comment by kennycason** [comment]
@ravisanwal Also pinging you here since you guys will likely still need this feature for your product. Also, check out this issue as there was a worry of ES removing the ability to add your own shard balancer that is now closed (https://github.com/simplymeasured/tempest/issues/34)

Also adding visibility on this repo to see if we even want to maintain it on SM, or migrate this to one of our personal GH's (mine or @aewhite 's)

**Comment by muffix** [comment]
I've been playing with the code and started a port to 5.x (and eventually later versions). I'm looking to open a PR to add the functionality soon.

**Comment by kennycason** [comment]
@muffix @luflow awesome. I want to say @aewhite has some work for 5.x but it's just lacking testing. I need to follow up with @ravisanwal and see the state of things internally. I have spoken with @aewhite and it's on his roadmap to work on as well.

**Comment by muffix** [comment]
As promised, I opened the PR above based on the changes by @aewhite on the `v2.2.2-ES5.6.6-SNAPSHOT` branch. No functional changes, but there were quite a few changes in the API and test setup. 

A slight caveat is that we can at the moment only support `5.6.10` as the latest versions because wer're missing releases for higher Elastic 5.6.x versions for a test dependency. See the PR for details. If we're not interested in running the tests, we can just bump the Elastic version in the `pom.xml`. Just to be sure, I asked for retrospective releases for the missing versions (https://github.com/codelibs/elasticsearch-cluster-runner/issues/31). 

I'll have a look if there are any further changes for Elastic 7.x and will raise a separate PR for that. We could also try to refactor the test setup (Elastic 5 now provides a lot of good stuff that we can use) so that we can potentially get rid of the dependency entirely.

**Comment by muffix** [comment]
@luflow I've got another branch for Elastic 7.5.2 based on the work above: https://github.com/muffix/tempest/tree/elastic7. Going to open a separate PR if/once we're happy with this change.

**Comment by luflow** [comment]
@muffix should we just create a new issue for the ES 7.x support? Just one thing I already saw: maybe we should set the plugin versions to the respecive ES support? In your branch the plugin version is v5 while the supported es version is 7.x - maybe its better to have a v7 plugin for 7.x just to reduce the confusion.

**Comment by muffix** [comment]
I think a separate issue for v7 support is a good idea. 

Tempest version numbers so far didn't really follow an Elastic version, I think. So the idea was to bump the major version with each major Elastic version change. So 5.x support becomes 3.x, 6.x support 4.x, 7.x support 5.x. 
I think it's still the same software, just built against a different target. 

I'm not too fussed about the version numbering. Multiple options there to choose from. They all work and I'm happy to go with whatever is personal preference of the owners.

**Comment by kennycason** [comment]
I don't think we'd be opposed to aligning version numberings as that may make things easier. Also just an fyi, I spoke to Andrew briefly so he is aware of this PR. Afterwards we can get @ravisanwal who is still at Sprout to merge in. I think it may be worthwhile confirming if the current GitHub page is the proper long term place for this plugin as well. (i.e. @aewhite would this plugin make more sense to be maintained from your account instead?)

**Comment by muffix** [comment]
@aewhite @kennycason Just wondering if you have an update on this?

**Comment by luflow** [comment]
@muffix @aewhite @kennycason I would also like to know how a test setup would look like? Is there any documentation for a build process & installing the built plugin into the cluster? Would love to try it out in the next days. We just finished our rolling upgrade to 7.6.0. I would just copy our Cluster onto my local machine and  try out the plugin.

**Comment by muffix** [comment]
@luflow The plugin registers a `ShardsAllocator` when installed. Then you can choose to use it by setting `cluster.routing.allocation.type` in the Elastic config to `tempest` (defaults to `balanced`, which is the built-in `BalancedShardsAllocator`). 

Here is where it's defined: https://github.com/simplymeasured/tempest/blob/de7e631241f3d41e959d4fc6208ea878f9122d4d/src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/TempestPlugin.kt#L106

**Comment by luflow** [comment]
@muffix just to understand the procedure: Can I just checkout the repo in the es7 branch and copy it to my test cluster? That sounds too easy :D Thats why I asked if there is any build process / packaging involved.

**Comment by muffix** [comment]
Yes, the [`elastic7`](https://github.com/muffix/tempest/tree/elastic7) branch should work @luflow. You will need to build it and copy the resulting plugin to your cluster and enable it in the config as described [here](https://github.com/simplymeasured/tempest/issues/24#issuecomment-596478684).

**Comment by luflow** [comment]
@muffix I know these are noob questions but thats why I asked: How can I build the plugin? Is there any documentation on this?

Thank you :)

**Comment by muffix** [comment]
No problem at all, the `pom.xml` hs all you need to set up your build. Just run `mvn package` from the root of the repo to run build and tests with Maven. You'll then find the jar file in the `target` subdirectory.
